### ISSUE TITLE
Support peek() for mpsc::Receiver

### ISSUE THREAD

**Issue by shade** [tokio-rs/tokio#6965]
**Is your feature request related to a problem? Please describe.**
I have an application that needs to wait for a value to show up in a channel within a certain time-frame before running some business logic.

**Describe the solution you'd like**
The ideal solution would be to have a `async peek()` functionality which is ready when a value has been sent but does not consume the value.

```
let (sender, receiver) = mpsc::channel(10);

timeout(receiver.peek(), TIMEOUT).await?;

// Hand off the receiver to 
business_logic_task(receiver);
```

**Describe alternatives you've considered**
Alternative 1. Is to await on `recv()` and write our business logic to work with the first channel value as a special case.

Alternative 2. Is to wrap the Receiver in a [ReceiverStream](https://docs.rs/tokio-stream/latest/tokio_stream/wrappers/struct.ReceiverStream.html), convert it into a [Peekable](https://docs.rs/tokio-stream/0.1.16/tokio_stream/adapters/struct.Peekable.html) stream and call peek on that.

My main gripe is that after the conversion to [Peekable](https://docs.rs/tokio-stream/0.1.16/tokio_stream/adapters/struct.Peekable.html), we cannot get the original receiver back as we have a lot of pre-built legacy code that depends on Receiver type and we will have to change this to be

**Comment by shade** [comment]
FWIW, I'm happy to write out the PR. But would like to know if there's any strong objection before doing so.

**Comment by Darksonn** [comment]
My initial reaction is that I don't love this feature request.

It sounds like you don't actually need `peek` to return the value?

**Comment by shade** [comment]
> It sounds like you don't actually need peek to return the value?

Indeed, we just want a notification that the receiver is ready to be `recv()`-d or `try_recv()`-d without consumption of any values.  Maybe `async fn ready()` would be a better?

Also, seems like some other folks have been asking for similar feature as well hence the peek idea:

- https://stackoverflow.com/questions/73405054/wait-for-an-mpsc-channel-to-have-messages-without-receiving-any-message
- https://github.com/crossbeam-rs/crossbeam-channel/issues/12 (though this is not async rust)
 

Would be curious to understand any objections you may have :)

**Comment by AlexvZyl** [comment]
Would also like to have something similar to this.  From my point of view this seems like something that should be in the API?  Unless misunderstand something, in which case I would love to be corrected.

**Comment by OldMansKid** [comment]
I also run into situations where I need to peek a receiver every now and then, for example, doing something lazily only when a value is available from the receiver. So I frequently use a wrapper like struct PeekableReceiver<T> { receiver: Receiver<T>, cache: Option<T>}. But not a big deal since the wrapper is easy to implement.
